code/iaas/model/src/main/java/io/cattle/platform/core/dao/impl/DynamicSchemaDaoImpl.java

```diff
@@ -1,34 +1,117 @@
 
+import io.cattle.platform.core.addon.DynamicSchemaWithRole;
+import io.cattle.platform.core.constants.CommonStatesConstants;
 import io.cattle.platform.core.dao.DynamicSchemaDao;
 import io.cattle.platform.core.model.DynamicSchema;
 import io.cattle.platform.core.model.Service;
 import io.cattle.platform.db.jooq.dao.impl.AbstractJooqDao;
+import io.cattle.platform.util.type.CollectionUtils;
+import io.github.ibuildthecloud.gdapi.exception.ValidationErrorException;
+import io.github.ibuildthecloud.gdapi.util.TypeUtils;
+import io.github.ibuildthecloud.gdapi.validation.ValidationErrorCodes;
 
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+
+import org.apache.commons.lang3.StringUtils;
+import org.jooq.Record;
+import org.jooq.SelectConditionStep;
+import org.jooq.exception.InvalidResultException;
 
 public class DynamicSchemaDaoImpl extends AbstractJooqDao implements DynamicSchemaDao {
 
     @Override
-    public List<? extends DynamicSchema> getSchemas(long accountId) {
-        return create().selectFrom(DYNAMIC_SCHEMA)
-                .where(DYNAMIC_SCHEMA.ACCOUNT_ID.eq(accountId))
+    public List<? extends DynamicSchema> getSchemas(long accountId, String role) {
+        List<Record> records = schemaQuery(accountId, role)
+                .orderBy(DYNAMIC_SCHEMA.CREATED.asc())
                 .fetch();
+        Map<String, List<Record>> nameSchemas = new HashMap<>();
+        for (Record record: records) {
+            if (nameSchemas.get(record.getValue(DYNAMIC_SCHEMA.NAME)) == null) {
+                nameSchemas.put(record.getValue(DYNAMIC_SCHEMA.NAME), new ArrayList<Record>());
+            }
+            nameSchemas.get(record.getValue(DYNAMIC_SCHEMA.NAME)).add(record);
+        }
+        List<DynamicSchema> recordsToReturn = new ArrayList<>();
+        for (List<Record> list: nameSchemas.values()) {
+            DynamicSchema schema = pickRecordOnPriority(list, accountId, role);
+            if (schema != null) {
+                recordsToReturn.add(schema);
+            }
+        }
+        return recordsToReturn;
     }
 
-    @Override
-    public DynamicSchema getSchema(String name, long accountId) {
-        return create().selectFrom(DYNAMIC_SCHEMA)
+    private SelectConditionStep<Record> schemaQuery(long accountId, String role) {
+        return create()
+                .select()
+                .from(DYNAMIC_SCHEMA).leftOuterJoin(DYNAMIC_SCHEMA_ROLE)
+                .on(DYNAMIC_SCHEMA_ROLE.DYNAMIC_SCHEMA_ID.eq(DYNAMIC_SCHEMA.ID))
                 .where(DYNAMIC_SCHEMA.ACCOUNT_ID.eq(accountId)
-                        .and(DYNAMIC_SCHEMA.NAME.eq(name)))
-                .orderBy(DYNAMIC_SCHEMA.CREATED.asc())
-                .fetchAny();
+                        .and(DYNAMIC_SCHEMA_ROLE.ROLE.eq(role))
+                        .and(DYNAMIC_SCHEMA.STATE.ne(CommonStatesConstants.PURGED)))
```

Comment:
We typically do REMOVED.isNotNull() for this type of check...any explicit reason you're just checking on purged and not considering removed?